### PR TITLE
Move pytorch distributed tests to separate folder for contbuild.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30445 Move pytorch distributed tests to separate folder for contbuild.**

Create distributed and rpc directories under caffe/test for better management
of unit tests.

For this purpose, we've created a new `torch.testing._internal` package where
we install all of our test libraries. This ensures we don't assume that all test files
are in the test/ folder (which enables them to use test libraries without imports)

Differential Revision: [D18702786](https://our.internmc.facebook.com/intern/diff/D18702786/)

### CODE REVIEW

**Comment by pritamdamania87** [comment]
Need to fix the OSS build.

**Comment by pietern** [comment]
Can we still run `python test/distributed/test_c10d.py` for example?

Paging @ezyang for test changes.

**Comment by mrshenli** [comment]
There was a [discussion](https://github.com/pytorch/pytorch/pull/29747#discussion_r346404506) on whether we should keep those tests in the top level dir.

**Comment by pritamdamania87** [comment]
> Can we still run python test/distributed/test_c10d.py for example?

This currently doesn't work without adding an appropriate `PYTHONPATH`. 

> There was a discussion on whether we should keep those tests in the top level dir.

I agree we should have a nice way to just run tests using `python test/distributed/test_c10d.py`, but I think we can achieve this without keeping everything in one directory. One way to do this would be to set the PYTHONPATH appropriately we setup conda. We need to split out the tests in different directories in order to fix the oncall issue.

**Comment by pritamdamania87** [comment]
Following the instructions here https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#macos-and-linux, I was able to configure my conda environment to export the appropriate `PYTHONPATH` when the env is activated and I was able to run: `python test/distributed/test_c10d.py`

**Comment by pritamdamania87** [comment]
@pietern @mrshenli @ezyang Would like to to have some consensus on this PR. Do you feel the proposal I've mentioned above would be reasonable? Or are there better ways to make this change seamless for developers?

**Comment by ezyang** [comment]
> Following the instructions here https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#macos-and-linux, I was able to configure my conda environment to export the appropriate PYTHONPATH when the env is activated and I was able to run: python test/distributed/test_c10d.py

This still requires developer intervention. You have to assume that the majority of developers aren't going to know to do this.

Since you seem to really, really want to have the distributed tests in a folder, let's think about a few options.

1. Add some boilerplate code to each test file which sets up a correct sys.path before doing imports (only when `__name__ == __main__`. This would "fix" the direct import case because each Python script would be responsible for fixing the sys.path before proceeding with the rest of the code)
2. Move all importable test modules into a true module namespace (e.g., `torch._test` or maybe `_torch_test` top level, or I suppose maybe even `test.blah`. Whatever you choose, you have to add a `__init__.py` file) and ensure that you import them in an explicitly qualified manner, e.g., no more `import test_common`, you have to say `import test.test_common`. Whatever the module namespace is, you have to "install" it in the same way the torch module is installed by `python setup.py develop`.

**Comment by pritamdamania87** [comment]
> Add some boilerplate code to each test file which sets up a correct sys.path before doing imports (only when __name__ == __main__. This would "fix" the direct import case because each Python script would be responsible for fixing the sys.path before proceeding with the rest of the code)

This might be tricky to maintain since we can easily forget to do this for new test files.

> Move all importable test modules into a true module namespace (e.g., torch._test or maybe _torch_test top level, or I suppose maybe even test.blah. Whatever you choose, you have to add a __init__.py file) and ensure that you import them in an explicitly qualified manner, e.g., no more import test_common, you have to say import test.test_common. Whatever the module namespace is, you have to "install" it in the same way the torch module is installed by python setup.py develop.

I thought about doing this, but wasn't sure if its fine to install test libraries this way. I'm assuming we would need to do this even when users run "python setup.py install" right? I'm assuming developers would expect to run tests even after the install command?

**Comment by ezyang** [comment]
> I thought about doing this, but wasn't sure if its fine to install test libraries this way. I'm assuming we would need to do this even when users run "python setup.py install" right? I'm assuming developers would expect to run tests even after the install command?

Yes, I think ordinarily `develop` and `install` do the same thing, except that develop symlinks to your source dir, while install copies it into site-packages. It is a little goofy to be distributing some extra Python code that no regular user would ever use, but an extra 3kb file isn't going to be noticed in the was of our 500mb binary distributions.

**Comment by pritamdamania87** [comment]
@ezyang Updated the PR based on the idea we discussed. Installed the test libraries to a package `torch.testlib`.

**Review by pietern — COMMENTED** [review]
`s/testlib/testing/`?

**Comment by pritamdamania87** [comment]
> s/testlib/testing/?

Isn't testlib more appropriate since these are actually test libraries. In the internal build we only install test libraries, for the OSS build I haven't figured out a way to exclude the actual test files (test_*)

**Comment by ezyang** [comment]
Well, on the subject of naming: (1) we already have `torch.testing`, and (2) the modules should be named in a way that makes it clear they are not for external use; usually that means putting an underscore in front of them or something

**Comment by pietern** [comment]
@ezyang Arguably there's quite a bit of functionality that should sit in `torch.testing` instead, as it is useful for testing for users as well. Since we already have it, we could reuse it, and move all the internal bits to torch.testing._internal or something. This could encourage folks to think twice about the public testing API instead of defaulting it to internal only helpers.

**Comment by pritamdamania87** [comment]
@ailzhang For the failing xla build. It's a simple fix in torch/xla since one of the patches don't apply.

**Comment by ailzhang** [comment]
Thanks Pritam! Yea we can fix it after this PR lands.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by pritamdamania87** [comment]
@ailzhang The fix is pretty simple and I have a patch: https://github.com/pytorch/xla/pull/1473

**Comment by ailzhang** [comment]
Thanks!

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit be5d85d:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=be5d85d8eb89788b1436e7b286ed4ed8e81508de).

### 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/3913365) pytorch_xla_linux_xenial_py3_6_clang7_build (1/1)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=72710205))
```
Dec 12 01:29:57 +++ eval 'extract_trap_cmd '
Dec 12 01:29:57 ++++ extract_trap_cmd
Dec 12 01:29:57 ++++ printf '%s\n' ''
Dec 12 01:29:57 +++ printf '%s\n' cleanup
Dec 12 01:29:57 ++ trap -- '
Dec 12 01:29:57 cleanup' EXIT
Dec 12 01:29:57 ++ which sccache
Dec 12 01:29:57 ++ sccache --stop-server
Dec 12 01:29:57 Stopping sccache server...
Dec 12 01:29:57 error: couldn't connect to server
Dec 12 01:29:57 caused by: Connection refused (os error 111)
Dec 12 01:29:57 ++ true
Dec 12 01:29:57 ++ rm /var/lib/jenkins/sccache_error.log
Dec 12 01:29:57 rm: cannot remove '/var/lib/jenkins/sccache_error.log': No such file or directory
Dec 12 01:29:57 ++ true
Dec 12 01:29:57 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log
Dec 12 01:29:57 ++ SCCACHE_IDLE_TIMEOUT=1200
Dec 12 01:29:57 ++ RUST_LOG=sccache::server=error
Dec 12 01:29:57 ++ sccache --start-server
Dec 12 01:29:57 Starting sccache server...
Dec 12 01:29:57 ++ sccache --zero-stats
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

**Comment by pritamdamania87** [comment]
@ezyang It looks like all builds are passing (the windows build is failing upstream and the xla build needs a fix on the xla side which I have a patch for: https://github.com/pytorch/xla/pull/1473). Did you have any other comments on this PR apart from renaming `testlib` to `_testlib`? Would like to address all comments at once and update this PR once since its grown quite a bit :)

File: setup.py [hunk 1]
```diff
@@ -766,6 +770,12 @@ def print_box(msg):
         packages=packages,
         entry_points=entry_points,
         install_requires=install_requires,
+        package_dir={
+            'torch.testlib' : './test',
+            'torch.testlib.data' : './test/data',
+            'torch.testlib.expect' : './test/expect',
+            'torch.testlib.distributed.rpc' : './test/distributed/rpc'
+        },
```
Comment (ezyang) [inline setup.py]: The use of `package_dir` for the entirety of the `test` directory looks very suspicious to me. I guess the reasoning is, you didn't want to move the `test` folder? Better say so...

File: test/common_utils.py [hunk 2]
```diff
@@ -1370,3 +1372,15 @@ def load_tests(loader, tests, pattern):
             check_test_defined_in_running_script(test)
             test_suite.addTest(test)
     return test_suite
+
+
+def _assertGradAndGradgradChecks(test_case, apply_fn, inputs):
+    # call assert function rather than returning a bool since it's nicer
+    # if we get whether this failed on the gradcheck or the gradgradcheck.
+    test_case.assertTrue(gradcheck(apply_fn, inputs))
+    test_case.assertTrue(gradgradcheck(apply_fn, inputs))
```
Comment (ezyang) [inline test/common_utils.py]: What's going on here?

File: test/distributed/test_data_parallel.py [hunk 3]
```diff
@@ -18,11 +24,11 @@ def __init__(self, t):
             def forward(self, x):
                 return x * self.t_rg + self.t_not_rg
 
-        m = TestModule(torch.randn(100, device='cuda', requires_grad=True))
+        m = TestModule(torch.randn(100, device='cuda', requires_grad=True, dtype=torch.double))
```
Comment (ezyang) [inline test/distributed/test_data_parallel.py]: Why did you have to add `dtype=double` here?

File: test/jit/test_class_type.py [hunk 4]
```diff
@@ -483,7 +483,7 @@ def forward(self, a):
 
         # classes are globally registered for now, so we need to clear the JIT
         # registry to simulate loading a new model
-        jit_utils.clear_class_registry()
+        torch.testlib.jit_utils.clear_class_registry()
```
Comment (ezyang) [inline test/jit/test_class_type.py]: Next time, just import `torch.testlib.jit_utils as jit_utils` so you don't have to fix the use sites... (no action needed on this PR)

File: test/run_test.py [hunk 5]
```diff
@@ -140,8 +154,16 @@ def run_test(executable, test_module, test_directory, options, *extra_unittest_a
     # Can't call `python -m unittest test_*` here because it doesn't run code
     # in `if __name__ == '__main__': `. So call `python test_*.py` instead.
     argv = [test_module + '.py'] + unittest_args + list(extra_unittest_args)
+
+    # Add 'test_directory' to PYTHONPATH so that imports work.
+    env = os.environ.copy()
+    if 'PYTHONPATH' in env:
+        env['PYTHONPATH'] = os.pathsep.join([env['PYTHONPATH'], test_directory])
+    else:
+        env['PYTHONPATH'] = test_directory
```
Comment (ezyang) [inline test/run_test.py]: I thought the point of all the discussing we did was so that you don't have to do these lines

**Comment by ezyang** [comment]
Diff looks mostly reasonable, but I expected just renaming of imports in the changes of the diff and there are substantive changes which are confusing to me. On the subject of renaming, let's put it under `torch.testing._internal` as suggested by @pietern

File: setup.py (hunk 1 above)
Comment (ezyang) [reply setup.py]: I'm sorry for changing my mind about this. I don't think it is a good idea to use the `package_dir` trick here. Just go ahead and move the files. Only the files that are imported will get moved, is that right? That will limit the blast radius of the move, since those files churn less than the real test files. What I am very afraid of is that people we see `import torch.testing._internal`, they will go to `torch/testing/_internal` and they will be like "Wut, where are the files?" I really don't want to have to explain this to every new person who comes and works on the codebase.

**Comment by pritamdamania87** [comment]
@ezyang Addressed your comments, renamed the package to `torch.testing._internal` and fixed most of the build issues. Could you take another look at this PR? Thanks!

File: test/distributed/test_data_parallel.py (hunk 3 above)
Comment (ezyang) [reply test/distributed/test_data_parallel.py]: This is probably related to the fact that we globally change the default precision in the tests. The precision is set like this in test files:

```
torch.set_default_dtype(torch.double)
```

It would be good to understand if this was somehow set prior to the move here (though I don't see how). You could check this by running `torch.get_default_dtype()` at this site and seeing what the dtype is.

File: test/run_test.py [hunk 6]
```diff
@@ -71,8 +72,21 @@
...
+# For tests not in the test/ directory.
+DIR_FOR_TESTS = {
+    "c10d": "distributed",
+    "c10d_spawn": "distributed",
+    "data_parallel": "distributed",
+    "distributed": "distributed",
+    "nccl": "distributed",
+    "rpc_spawn": "distributed/rpc",
+    "dist_autograd_spawn": "distributed/rpc",
+    "dist_optimizer_spawn": "distributed/rpc",
```
Comment (ezyang) [inline test/run_test.py]: This is definitely not how you would have written the code to start with ;)

File: torch/testing/_internal/common_utils.py [hunk 7]
```diff
@@ -1029,14 +1031,10 @@ def remove_prefix(text, prefix):
...
-        # NB: we take __file__ from the module that defined the test
-        # class, so we place the expect directory where the test script
-        # lives, NOT where test/common_utils.py lives.  This doesn't matter in
-        # PyTorch where all test scripts are in the same directory as
-        # test/common_utils.py, but it matters in onnx-pytorch
+        # Always look for 'expect' in the same directory as 'common_utils.py'.
         module_id = self.__class__.__module__
         munged_id = remove_prefix(self.id(), module_id + ".")
-        test_file = os.path.realpath(sys.modules[module_id].__file__)
+        test_file = os.path.realpath(__file__)
```
Comment (ezyang) [inline torch/testing/_internal/common_utils.py]: This change doesn't make sense to me. In particular, I don't see any justification why you haven't just broken `onnx-pytorch` by changing it this way. You should revert this change and just move the expect files to the expected directories (now that you have reorganized the folder structure.)

**Review by ezyang — CHANGES_REQUESTED** [review]
The expect tests should not have been moved.

File: test/run_test.py (hunk 6 above)
Comment (pritamdamania87) [reply test/run_test.py]: Will fix this to find the appropriate directory automatically.

File: torch/testing/_internal/common_utils.py (hunk 7 above)
Comment (pritamdamania87) [reply torch/testing/_internal/common_utils.py]: The `expect` directory moved to `torch/testing/_internal/expect` and the old code here no longer worked correctly since it looked for the files in the wrong directory. This codepath is used by pytorch tests and also onnx tests. Although onnx tests had a separate expect directory: `test/onnx/expect`. Instead of having separate expect folders, I merged the into one (`torch/testing/_internal/expect`) and this common component always looks in that directory for the expect files.

I feel this more consistent than having separate `expect` directories and handling the `onnx` codepath separately here.

Comment (ezyang) [reply torch/testing/_internal/common_utils.py]: > Instead of having separate expect folders, I merged the into one (torch/testing/_internal/expect) and this common component always looks in that directory for the expect files.

Here's is my reasoning why there should be separate expect folders. Let me suppose that I'm adding an expect test. Typically, I'll be editing a file in `test/`, like `test/test_jit.py`. A principle that I generally like to see is that related files should be close by; so if I need to update expect tests, they should be similarly close by, like `test/expect/...`. Suppose that I had an out-of-tree test suite, like `pytorch-onnx/test`; if the test directory is based on the folder location, I can put the expect tests out-of-tree too, and not have to put them in `torch/testing/_internal/expect`. The consistency is that expect tests live with the test script which generates them!

Comment (pritamdamania87) [reply torch/testing/_internal/common_utils.py]: I see your point. I moved all the expect files back to the original location and nuked `torch.testing._internal.expect`.

File: test/run_test.py [hunk 8]
```diff
@@ -412,6 +416,16 @@ def get_selected_tests(options):
...
 
+def find_test_file(test_name, test_directory):
+    test_file_name = '{}.py'.format(test_name)
+    for root, dirs, files in os.walk(test_directory):
+        for file in files:
+            if file == test_file_name:
+                abspath = os.path.join(root, file)
+                return os.path.relpath(abspath, test_directory)
+
+    raise Exception('Could not find test file for: {}'.format(test_name))
```
Comment (ezyang) [inline test/run_test.py]: Hmm, it's my bad for not being more explicit. When I mentioned that the previous code was not "how I would have written from scratch", what I had meant is that if I were writing this from scratch, I would have referred to tests by qualified name, so e.g., `jit/test_builtins` rather than just by the basename `test_builtins`. There are a few reasons why this makes more sense to me than searching for a file with this name (this patch) or hard-coding the path associated with the name (the previous patch), but the main reason is, usually when people introduce a directory structure, they imagine that it's OK for a file with the same name to live in different directories; e.g., `jit/test_basic` and `distributed/test_basic`. Both the old code, and this new code, violate that assumption: you must name every test uniquely, which could lead to weirdly redundant names like `jit/test_jit_basic`

File: test/test_tensorboard.py [hunk 9]
```diff
@@ -460,13 +460,21 @@ def test_mesh(self):
...
     functionName = function_ptr.id().split('.')[-1]
-    expected_file = os.path.join(test_dir,
-                                 "expect",
-                                 'TestTensorBoard.' + functionName + ".expect")
+    return os.path.join(test_dir,
+                        "expect",
+                        'TestTensorBoard.' + functionName + ".expect")
+
+def read_expected_content(function_ptr):
+    expected_file = get_expected_file(function_ptr)
```
Comment (ezyang) [inline test/test_tensorboard.py]: You don't have to fix this, but this is really weird. I don't know why `test_tensorboard` has another copy of the "find expect file" logic...

**Review by ezyang — APPROVED** [review]
Alright, sorry about bikeshedding so much. Have the stamp. Take a look over my comments and see if you agree. If you want to just get this in, that's fine by me too.

File: test/test_tensorboard.py (hunk 9 above)
Comment (pritamdamania87) [reply test/test_tensorboard.py]: The logic in common_utils.py is slightly different and it actually uses the `subname` parameter to determine the right file name.

File: test/run_test.py (hunk 8 above)
Comment (pritamdamania87) [reply test/run_test.py]: Let me address this in a follow up PR. We can add this functionality, although one question I had was what user behavior we would like to have for run_test.py. Currently you can run specific tests as follows: `run_test.py -i jit` or `run_test.py -i dist_optimizer_spawn`. Do we want to change this to be: `run_test.py -i test_jit` and `run_test.py -i distributed/rpc/test_dist_optimizer_spawn` (this option would modify existing behavior of run_test.py and I'm not sure if that breaks anything else). Or do we want something like this: `run_test.py -i jit` and `run_test.py -i distributed/rpc/dist_optimizer_spawn` ?

Comment (pritamdamania87) [reply test/run_test.py]: Actually, looks like we need to fix this in the current PR itself since we have the following:

```
test/jit/test_data_parallel.py
test/distributed/test_data_parallel.py (this test wasn't enabled in CI, but is now with this PR)
```
I can update the PR based on which approach mentioned in my previous comment makes sense to you.

Comment (ezyang) [reply test/run_test.py]: I don't know. Honestly, I have always hated the fact that we needed a `run_test.py` script at all; why can't I just run the test files directly! (There are reasons why this can't be conveniently done.) There seem to be two conflicting design goals:

* I don't want to change the API for `run_test.py`, because if you do you're going to affect people who run tests this way (e.g., you're going to have to change the CI scripts at the very least)
* I don't want complicated name mangling logic for translating inputs in the test script (since you need to split the path, strip off `test_` basename, and then put it back together)

I think, on balance, I would rather keep `run_test.py` simple (no mangling logic) and force people to use the new API. I suppose you could add some BC logic (e.g., if you can't find the test, prepend `test_` again) to keep the old API working. I would like it even *more* if I could just get rid of `run_test.py` entirely for test scripts that don't need it.

File: test/test_tensorboard.py (hunk 9 above)
Comment (ezyang) [reply test/test_tensorboard.py]: That's the story of life with copy pasted code, right :) Maybe if you could file a bug on this and CC the original code author that would be a big help

**Comment by pritamdamania87** [comment]
@ezyang Could you take another look at this PR? I modified run_test.py and forced the new API. As part of this change I also fixed all the CI scripts I could find to use the new API. I looked into adding some backward compatibility behavior, but doing this was a little messy. The reason is that the test choices are validated by the args parser and it looks though the `TESTS` array to validate if the test name provided is valid. I would have to remove this validation from the arg parser and do it myself in `run_test.py` to appropriately fallback to the old behavior. As a result, I thought it would be better to just force the new API.

**Review by ezyang — APPROVED** [review]
OK, looks good. Make sure you post an announce on devs saying that `run_test.py` api has changed

File: test/distributed/test_data_parallel.py (hunk 3 above)
Comment (gchanan) [reply test/distributed/test_data_parallel.py]: It would have been good to follow @ezyang's suggestion here and do some digging (or just set the default dtype).

As it is now, there is a mixture of: 1) functions that set the dtype explicitly to double (I'm guessing these are ones you changed because the precision issues were causing failures, 2) tests that don't set a dtype (I'm guessing these are meant to be double but don't happen to cause precision issues), 3) tests that set dtype to float explicitly (these are pointless now), 4) tests that run over some dtypes (these are fine).

This isn't the worse, but it goes from something that was probably thought out to something that wasn't.

File: torch/testing/_internal/common_utils.py [hunk 10]
```diff
@@ -1443,3 +1445,16 @@ def load_tests(loader, tests, pattern):
...
+
+
+def _assertGradAndGradgradChecks(test_case, apply_fn, inputs):
+    # call assert function rather than returning a bool since it's nicer
+    # if we get whether this failed on the gradcheck or the gradgradcheck.
+    test_case.assertTrue(gradcheck(apply_fn, inputs))
+    test_case.assertTrue(gradgradcheck(apply_fn, inputs))
+
+
+dtype2prec = {torch.float: 1e-5,
```
Comment (gchanan) [inline torch/testing/_internal/common_utils.py]: we _really_ shouldn't be canonicalizing this.

1) The name is really bad (it's _maybe_ good enough for some one-off tests in test_nn, but not as a common utility)
2) The numbers were selected empirically for some backwards cases, there's nothing special about them.

So we don't have to restructure the tests, can we rename this something like "dtype2prec_DONTUSE" and write a comment here about the recommended way of doing this (use `@precisionOverride` or just find a reasonable number for your use case).
